tokio/src/scope/cancellation_token.rs

```diff
@@ -0,0 +1,1186 @@
+                    // Cancellation was signalled. Since this cancellation signal
+                    // is set inside the Mutex, the old waiter must already have
+                    // been removed from the waiting list
+                    debug_assert_eq!(PollState::Done, wait_node.state);
+                    wait_node.task = None;
+                    Poll::Ready(())
+                } else {
+                    // The WaitForCancellationFuture is already in the queue.
+                    // The CancellationToken can't have been cancelled,
+                    // since this would change the is_cancelled flag inside the mutex.
+                    // Therefore we just have to update the Waker. A follow-up
+                    // cancellation will always use the new waker.
+                    wait_node.task = Some(cx.waker().clone());
+                    Poll::Pending
+                }
+            } else {
+                // Do nothing. If the token gets cancelled, this task will get
+                // woken again and can fetch the cancellation.
+                Poll::Pending
+            }
+        }
+    }
+
+    fn unregister(&self, wait_node: &mut ListNode<WaitQueueEntry>) {
+        debug_assert!(
+            wait_node.task.is_some(),
+            "waiter can not be active without task"
+        );
+
+        let mut guard = self.synchronized.lock().unwrap();
+        // WaitForCancellationFuture only needs to get removed if it has been added to
+        // the wait queue of the CancellationToken.
+        // This has happened in the PollState::Waiting case.
+        if let PollState::Waiting = wait_node.state {
+            // Safety: Due to the state, we know that the node must be part
+            // of the waiter list
+            if !unsafe { guard.waiters.remove(wait_node) } {
+                // Panic if the address isn't found. This can only happen if the contract was
+                // violated, e.g. the WaitQueueEntry got moved after the initial poll.
+                panic!("Future could not be removed from wait queue");
+            }
+            wait_node.state = PollState::Done;
+        }
+        wait_node.task = None;
+    }
+}
+
+/// A token which can be used to signal a cancellation request to one or more
+/// tasks.
+///
+/// Tasks can call [`CancellationToken::wait_for_cancellation`] in order to
+/// obtain a Future which will be resolved when cancellation is requested.
+///
+/// Cancellation can be requested through the [`CancellationToken::cancel`] method.
+///
+/// # Examples
+///
+/// ```ignore
+/// use tokio::select;
+/// use tokio::scope::CancellationToken;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let token = CancellationToken::new();
+///     let cloned_token = token.clone();
+///
+///     let join_handle = tokio::spawn(async move {
+///         // Wait for either cancellation or a very long time
+///         select! {
+///             _ = cloned_token.wait_for_cancellation() => {
```

Comment:
API design nit: Could the `CancellationToken`  implement `std::future::Future` itself? I think this would make the API feel more implace with other recievers like [`oneshot::Reciever`](https://docs.rs/tokio/0.2.11/tokio/sync/oneshot/struct.Receiver.html#impl-Future).

(It would also remove the need for a longer method like `wait_for_cancellation` which doesn't _feel_ like idiomatic Rust.

Reply:
No, because there must be a pinned Future, which stores the waiting node. If CancellationToken would embed it, it would need to be pinned too - which makes it kind of unusable.

What could however remove the future type and just create an `async fn wait_for_cancellation(&self)`.

Reply:
We would need `IntoFuture`.

Until then, the pattern I've been using is have the `fn` named as the end state, so:

```rust
token.cancelled().await
```

Reply:
Sounds good to me